Closest points algorithm with new strategies for spherical and geographic geometries except boxes

vissarion opened the pull request:
This PR build on top of https://github.com/boostorg/geometry/pull/933

vissarion commented:
This PR includes documentation for the `closest_points` algorithm. Even if the documentation is build without errors (both locally and on CI) the generated algorithm is empty (i.e. it appears in the html list of algorithms but the algorithm's html page only contains its name). Also, the `doc/generated/closest_points.qbk` is almost empty, it contains only the copyright info and comments ` [/ Generated by doxygen_xml2qbk...]`. 

Note that I followed the instructions from `https://github.com/boostorg/geometry/blob/develop/doc/readme.txt`, I am using `gcc-9` and `python 3.8` in `ubuntu 20.04.3 LTS`. 

Interestingly the branch https://github.com/boostorg/geometry/pull/707 generates the correct documentation. 

I am missing something here that I have to investigate more or is it a more general problem with the doc system?

barendgehrels commented on doc/src/examples/examples_utils/create_svg_closest_points.hpp:
```diff
@@ -0,0 +1,66 @@
+#define CREATE_SVG_CLOSEST_POINTS_HPP
+
+#include <fstream>
+#if defined(HAVE_SVG)
```
Very minor, using this define can go here :-)

vissarion replied:
sorry, I don't get it

awulkiew replied:
Probably that you can remove the `#ifdef`. Note also that since below `boost::geometry::svg_mapper` is created technically this `#include` is needed anyway. Unless you put the whole code behind the `#ifdef`. But AFAIU is some helper utility so there is no need for that.

barendgehrels reviewed (approved):
Thanks!

vissarion commented:
> Thanks!

Thanks for the review. Since this PR is on top of https://github.com/boostorg/geometry/pull/933 you have also reviewed and approved https://github.com/boostorg/geometry/pull/933 implicitly. 

@awulkiew it is up to you to review those PRs all in once or separately, e.g. if you have many comments it would be useful to separate otherwise not.

awulkiew commented on doc/doxy/Doxyfile:
```diff
@@ -1,187 +1,919 @@
-# Doxyfile 1.5.5
+# Doxyfile 1.8.17
+
+# This file describes the settings to be used by the documentation system
+# doxygen (www.doxygen.org) for a project.
+#
+# All text after a double hash (##) is considered a comment and is placed in
+# front of the TAG it is preceding.
+#
+# All text after a single hash (#) is considered a comment and will be ignored.
+# The format is:
+# TAG = value [value, ...]
+# For lists, items can also be appended using:
+# TAG += value [value, ...]
+# Values that contain spaces should be placed between quotes (\" \").
```
There are many additional comments in this file. Is this intentional?

vissarion replied:
This is automatic, this file is generated after upgrading to Doxyfile version 1.8.17.
The additions are almost all comments (probably manually removed from previous versions ?) but there are some extra options added and some old ones removed.

awulkiew replied:
Do we really need these additional comments? The file is long enought without them.

vissarion replied:
We do not need them but could be useful as a documentation. Also to my understanding one has to remove extra comments manually after every Doxyfile upgrade and that could be error prone. But I am OK removing those if you insist because I guess a similar decision had already been taken in the past.

awulkiew replied:
> remove extra comments manually after every Doxyfile upgrade

I thought you made this on purpose. AFAIU `doxygen` can generate a warning if something is wrong with the input file but will not modify it unless you pass `-u` option. This option is not passed in `make_qbk.py` so you must have passed it by yourself.

For me doxygen 1.9.1 reports:
```
warning: Tag 'TCL_SUBST' at line 39 of file 'Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 172 of file 'Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'PERL_PATH' at line 295 of file 'Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'MSCGEN_PATH' at line 300 of file 'Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
```

So I'd simply remove these variables from the file if I wanted to silence this warning or leave the doxyfile unmodified, as it is now. Because this is not related to `closest_points` in any way.

awulkiew commented on doc/index/Doxyfile:
```diff
@@ -1,17 +1,131 @@
+# For lists, items can also be appended using:
+# TAG += value [value, ...]
+# Values that contain spaces should be placed between quotes (\" \").
+
```
The modifications in this file are not needed. Or am I missing something?

vissarion replied:
This is again automatic, this file is generated after upgrading to Doxyfile version 1.8.17.
The additions are almost all comments (probably manually removed from previous versions ?) but there are some extra options added and some old ones removed. 
For example the [`PERL_PATH`](https://github.com/boostorg/geometry/blob/develop/doc/index/Doxyfile#L295) which seems to be deprecated; see https://github.com/doxygen/doxygen/commit/6d1535c38fe6bdaa2a00fff0e7e43774a740a4ce

awulkiew commented on doc/readme.txt:
```diff
@@ -31,7 +31,8 @@ reference: generated documentation QBK files (by doxygen_xml2qbk)
 Per new algorithm (e.g. foo), one should add:
-1) in file boost/geometry/algorithms/foo.hpp, include a "\ingroup foo" in the doxygen comments
+1) in file boost/geometry/algorithms/foo.hpp, include a "\ingroup foo" in the doxygen comments 
+   (if a directory is created e.g. boost/geometry/algorithms/detail/foo the path should be added in doc/doxy/Doxyfile)
```
+1 though this is a different issue than the one mentioned in 1) Shouldn't it be a separate point, e.g. 3) ?

vissarion replied:
OK I added as a separate point, 2) and update the enumeration.

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_or_areal_to_areal.hpp:
```diff
@@ -0,0 +1,269 @@
+
+        using point_type = typename std::conditional
+        <
+            std::is_same<coordinate_type<Linear>, most_precise_type>::value,
```
`geometry::coordinate_type` is not a type alias so here it should be `typename coordinate_type<Linear>::type`.

Side note. We could think about adding type aliases and variable templates for traits: `coordinate_t`, `point_t`, `dimension_v`, etc.

vissarion replied:
> `geometry::coordinate_type` is not a type alias so here it should've been `typename coordinate_type<Linear>::type`.

sure, thanks for noticing 

> Side note. We could think about adding type aliases and variable templates for traits: `coordinate_t`, `point_t`, `dimension_v`, etc.

+1

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_or_areal_to_areal.hpp:
```diff
@@ -0,0 +1,269 @@
+
+        using linestring_type = geometry::model::linestring<point_type>;
+        
+        /* TODO: currently intersection does not support tupled input
+         *       this should be implemented directly with dynamic geometries
+        using polygon_type = geometry::model::polygon<point_type>;
+        std::tuple
+        <
+            geometry::model::multi_point<point_type>, 
+            geometry::model::multi_linestring<linestring_type>,
+            geometry::model::multi_polygon<polygon_type>
+        > tp;
+        bool intersect_tp = geometry::intersection(linear, areal, tp, strategies);
+        */
```
It should be possible to pass tupled geometry as output geometry of `intersection()`.
https://github.com/boostorg/geometry/blob/develop/test/algorithms/set_operations/intersection/intersection_tupled.cpp

vissarion replied:
sure, but some cases are not supported, e.g. linestring - multipolygon, I updated my comment to explain that.

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_or_areal_to_areal.hpp:
```diff
@@ -0,0 +1,269 @@
+
+        using point_type = typename std::conditional
+        <
+            std::is_same<coordinate_type<Areal1>, most_precise_type>::value,
```
`typename coordinate_type<Areal1>::type`

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_or_areal_to_areal.hpp:
```diff
@@ -0,0 +1,269 @@
+        
+        if (! boost::empty(mp_out))
+        {
+            set_segment_from_points::apply(*boost::begin(mp_out), 
```
In case you wanted to make this shorter we have `util::front()`.

vissarion replied:
sure, but I do not want to assert on empty `mp_out`

awulkiew replied:
`mp_out` is checked for emptiness above so the assertion will not fail. Or do you mean that you don't want to perform an additional unnecessary check? Anyway I'm ok with leaving it as it is.

vissarion replied:
> Or do you mean that you don't want to perform an additional unnecessary check?

yes

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/multipoint_to_geometry.hpp:
```diff
@@ -0,0 +1,326 @@
+            <
+                typename point_type<Segment>::type
+            >;
+        linestring_type linestring;
+        convert(segment, linestring);
+        multipoint_to_linear::apply(multipoint, linestring, shortest_seg, strategies);
```
Converting segment to linestring seems to be unnecessary but if it simplifies the implementation then I'm fine with it.

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_to_linear.hpp:
```diff
@@ -0,0 +1,155 @@
+
+        if (geometry::num_segments(linear1) < geometry::num_segments(linear2))
+        {
+            point_or_segment_range_to_geometry_rtree::apply(
```
This can be called for empty geometry which will cause assertion failure inside `point_or_segment_range_to_geometry_rtree`. Or am I missing something?

vissarion replied:
Empty input is checked by the interface here: https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/closest_points/interface.hpp#L178
Several empty input combinations are also tested in unit tests.

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/linear_to_linear.hpp:
```diff
@@ -0,0 +1,155 @@
+            return;
+        }
+
+        point_or_segment_range_to_geometry_rtree::apply(
```
Here also empty geometry may be passed and `point_or_segment_range_to_geometry_rtree` is unable to handle that.

vissarion replied:
Empty input is checked by the interface here: https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/closest_points/interface.hpp#L178

awulkiew commented on include/boost/geometry/algorithms/detail/closest_points/segment_to_segment.hpp:
```diff
@@ -0,0 +1,154 @@
+        closest_points::creturn_t<Segment1, Segment2, Strategies> d[4];
+        
+        //TODO: this could be simplified by getting distance strategy in a unique way
+        d[0] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(cp0, q[0]))
+            >::apply(strategies.distance(cp0, q[0])).apply(cp0, q[0]);
+        d[1] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(cp1, q[1]))
+            >::apply(strategies.distance(cp1, q[1])).apply(cp1, q[1]);
+        d[2] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(p[0], cp2))
+            >::apply(strategies.distance(p[0], cp2)).apply(p[0], cp2);
+        d[3] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(p[1], cp3))
+            >::apply(strategies.distance(p[1], cp3)).apply(p[1], cp3);
```
All of these calculate point to point distances right?

If that's the case you only need to create one strategy, e.g. using the first pair of points. But there is actually no need to pass the actual points. You could pass `detail::dummy_point`s [1]. This way you could create the strategy somewhere above where the points are unavailable. Furthermore instead of `get_comparable` you could use `strategies::distance::detail::make_comparable()` [2] to make umbrella strategy returning comparable legacy strategies from `distance()` member function. Though this would be an additional dependency so you might want to choose to stick with `get_comparable` but using it only once. Anyway, this is how this code could be reimplemented:
```
auto const cds = strategies::distance::detail::make_comparable(strategies)
                     .distance(detail::dummy_point(), detail::dummy_point());
d[0] = cds.apply(cp0, q[0]);
d[1] = cds.apply(cp1, q[1]);
// etc.
```

[1] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/dummy_geometries.hpp#L26
[2] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/distance/comparable.hpp#L51

awulkiew commented on include/boost/geometry/strategies/closest_points/geographic.hpp:
```diff
@@ -0,0 +1,121 @@
+    typename Spheroid = srs::spheroid<double>,
+    typename CalculationType = void
+>
+class geographic
```
Out of curiosity, why didn't you just define `closest_points()` member function here?

vissarion replied:
The definition of base class in detail namespace is not needed, probably leftover from some previous implementation. Thanks for noticing.

awulkiew commented on include/boost/geometry/strategies/closest_points/spherical.hpp:
```diff
@@ -0,0 +1,112 @@
+    typename RadiusTypeOrSphere = double,
+    typename CalculationType = void
+>
+class spherical
```
Here as well, why is base class in `detail` needed instead of defining membre functions here?

vissarion replied:
The definition of base class in detail namespace is not needed.

awulkiew commented on include/boost/geometry/strategies/spherical/distance_cross_track.hpp:
```diff
@@ -508,22 +549,22 @@ template
 class cross_track
 {
 public :
-    typedef within::spherical_point_point equals_point_point_strategy_type;
+    using equals_point_point_strategy_type = within::spherical_point_point;
 
-    typedef intersection::spherical_segments
+    using relate_segment_segment_strategy_type = intersection::spherical_segments
         <
             CalculationType
-        > relate_segment_segment_strategy_type;
+        >;
 
     static inline relate_segment_segment_strategy_type get_relate_segment_segment_strategy()
     {
         return relate_segment_segment_strategy_type();
     }
 
-    typedef within::spherical_winding
+    using point_in_geometry_strategy_type = within::spherical_winding
         <
             void, void, CalculationType
-        > point_in_geometry_strategy_type;
+        >;
```
These are probably not used at all after the introduction of umbrella strategies.

awulkiew commented on include/boost/geometry/strategies/geographic/closest_points_pt_seg.hpp:
```diff
@@ -0,0 +1,121 @@
+                                        get_as_radian<0>(p), get_as_radian<1>(p),
+                                        base_t::m_spheroid);
+
+            model::point
+                <
+                    typename base_t::template return_type<Point, PointOfSegment>::type,
+                    dimension<PointOfSegment>::value,
+                    typename coordinate_system<PointOfSegment>::type
+                > cp;
+            
+            geometry::set_from_radian<0>(cp, result.lon);
+            geometry::set_from_radian<1>(cp, result.lat);
+
+            return cp;
```
So the result is a point using the coordinate type being the result of distance? If that's the case here then for consistency it has to be the case in cartesian CS too. So technically it's possible to pass points with integral coordinates and get a point with double coordinates as the output. Is that what we want?

vissarion replied:
No, this should be just like in spherical. Changed.

vissarion replied:
Rethinking of it, we probably should always promote to floating point since even in the case of input integral coordinates the closest point on the segment is possibly not be able to be represented by integral coordinates. This should happen in all CS.

awulkiew replied:
Yes, it makes sense. I don't know if this is correct since a user will typically pass the same point type as the in the input and output. Still internally an FP type can be used for calculation purposes even if at the end it will be converted to user-defined type.

If you take the coordinate type from the distance result in practice it is going to be FP in all coordinate systems.

vissarion replied:
OK, I left it as is and added a unit test with integral coordinates.

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -62,9 +62,37 @@ namespace boost { namespace geometry
 namespace detail
 {
+
+template <bool>
```
I propose to add a tparam name here to indicate what's happening. This would be `EnableClosestPoints` right?

Btw, if this utility alwasy has to take distance and coordinates of the closest point then it means that all of these always exist. If they exist they has to be calculated. So what is the purpose of assigning them conditionally if they are always there anyway? Or is this only the case in geographic CS?

vissarion replied:
Exactly, in geographic (unlike in spherical) computing distance and closest points is the same computation so `EnableClosestPoints` is only to avoid some (few) extra assignments. I do not mind removing it if you prefer.

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -254,17 +273,22 @@ private :
 #endif
-            if (inverse_distance_type::apply(lon3, lat3, pl_lon, pl_lat, spheroid).distance
-                < inverse_distance_type::apply(lon3, lat3, pr_lon, pr_lat, spheroid).distance)
+            CT const dist_l = inverse_distance_type::apply
+                (lon3, lat3, pl_lon, pl_lat, spheroid).distance;
+            CT const dist_r = inverse_distance_type::apply
+                (lon3, lat3, pr_lon, pr_lat, spheroid).distance;
```
Accroding to our guidelines lines can have up to 100 characters. I don't know if this would fit into 100 characters but in general putting parts of the expression in multiple lines makes it harder to read.

barendgehrels replied:
The official boost line length is 80 characters. I don't remember that we changed it.

I would actually be in favor of making it larger - but then maybe even more than 100 (for example 120). Years ago Bruno objected to make it larger than 80 characters, because he sometimes edits sources in a terminal. That might still be valid for some users. Also for showing a side-by-side diff the max line length is important.

But most probably for this one, we're just still restricted to Boost Guideline which is 80.
https://www.boost.org/development/requirements.html

awulkiew replied:
@barendgehrels from https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers

> The preferred line length is 80 characters, with maximum length of 100.

It's like that since guidelines were added to the [Wiki in 2014](https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers/9d366b7a52005b3b995e79501c07adeb5d0d75a5). And I'm pretty sure that we talked about this then.

I'm ok with increasing it to 120.

Isn't line length mainly defined by the size of the screen? Nowadays we have bigger screens or multiple screens. This also affects the size of the terminal as well as widths of diff columns.

awulkiew replied:
It's in the original version from the docs, since 2013:
https://github.com/boostorg/geometry/commit/10dd010ecc189276de1584187de847c4b4239e3c#diff-f7aee8e109b4ecd10dec03a3010cc6c110436140b709fa3632a5053c57d1bb84

vissarion replied:
> According to our guidelines lines can have up to 100 characters. I don't know if this would fit into 100 characters but in general putting parts of the expression in multiple lines makes it harder to read.

It doesn't fit. 
Is 
```
CT const dist_l = 
    inverse_distance_type::apply(lon3, lat3, pl_lon, pl_lat, spheroid).distance;
CT const dist_r = 
    inverse_distance_type::apply(lon3, lat3, pr_lon, pr_lat, spheroid).distance;
```
preferable?

barendgehrels replied:
> @barendgehrels from https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers
> 
> > The preferred line length is 80 characters, with maximum length of 100.
> 
> It's like that since guidelines were added to the [Wiki in 2014](https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers/9d366b7a52005b3b995e79501c07adeb5d0d75a5). And I'm pretty sure that we talked about this then.

:+1: thanks for the reminder!

So I'm already setting the right margin in the IDE to 80. Anyway: "preferred length is 80" but you may exceed it is actually not such a clear guideline... But anyway, I think we all do.

> 
> I'm ok with increasing it to 120.
> 
> Isn't line length mainly defined by the size of the screen? Nowadays we have bigger screens or multiple screens. This also affects the size of the terminal as well as widths of diff columns.

Sorry, I've to retract my own statement about 120, it's probably too much. Yes, the length is of course determined by the screen width - but also by the used font. I just did a `meld` (side by side view) in some adapted code and for me 107 is the personal limit. And of course for others the limit might be higher, but also smaller, all depending on personal settings.

So maybe better to just keep it as it is written there.

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -103,128 +144,106 @@ class geographic_cross_track
-        typedef typename FormulaPolicy::template inverse<CT, true, false, false, false, false>
-                inverse_distance_type;
+        using direct_distance_type = typename FormulaPolicy::template direct
+            <
+                CT,
+                true,
+                false,
+                false,
+                false
+            >;
```
I propose to put all of this in one line. I can't see any benefits of formatting it like this.

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -307,17 +337,18 @@ private :
-        typedef typename FormulaPolicy::template direct<CT, true, false, false, false>
-                direct_distance_type;
+        using inverse_distance_azimuth_quantities_type = typename 
+            FormulaPolicy::template inverse<CT, true, true, false, true, true>;
```
So here the same thing is formatted differently.

vissarion replied:
changed to 
```
using inverse_distance_azimuth_quantities_type = 
            typename FormulaPolicy::template inverse<CT, true, true, false, true, true>;
    
using inverse_dist_azimuth_type = 
            typename FormulaPolicy::template inverse<CT, false, true, false, false, false>;

using direct_distance_type = 
            typename FormulaPolicy::template direct<CT, true, false, false, false>;
```

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -328,22 +359,25 @@ private :
-            res34 = inverse_distance_azimuth_quantities_type::apply(res14.lon2, res14.lat2,
-                                                                    lon3, lat3, spheroid);
+            CT a4 = inverse_dist_azimuth_type
+                    ::apply(res14.lon2, res14.lat2, lon2, lat2, spheroid).azimuth;
```
I can see three types of formatting for such call in this file. In addition to this one there are also:
```
            CT const dist_l = inverse_distance_type::apply
                (lon3, lat3, pl_lon, pl_lat, spheroid).distance;
```
and
```
                s14 -= inverse_distance_type::apply(res14.lon2, res14.lat2,
                                                    pl_lon, pl_lat,
                                                    spheroid).distance/2;
```

I don't know which one is the best but we could pick one of them preferably with the least number of lines. Do we have guidelines for this?

I also insist on not breaking lines shorter than 101 characters.

vissarion replied:
what about this ?
```
s14 -= inverse_distance_type::apply(
    res14.lon2, res14.lat2, pl_lon, pl_lat, spheroid).distance/2;
```

vissarion replied:
for the rest see comments above

awulkiew replied:
Whatever works for you, just be consistent. I prefer the second one with all arguments at the same place but I'm ok with a different one if you prefer it.

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -431,21 +486,59 @@ private :
+                             CT const& lo3, CT const& la3, //query point p3
+                             Spheroid const& spheroid)
+    {
+        using inverse_dist_azimuth_type = typename FormulaPolicy::template 
+            inverse<CT, true, true, false, false, false>;
+
+        using inverse_dist_azimuth_reverse_type = typename FormulaPolicy::
+            template inverse<CT, true, true, true, false, false>;
```
Should we pick one formatting here or is this ok?

vissarion replied:
what about 
```
using inverse_dist_azimuth_type = 
            typename FormulaPolicy::template inverse<CT, true, true, false, false, false>;

using inverse_dist_azimuth_reverse_type = 
            typename FormulaPolicy::template inverse<CT, true, true, true, false, false>;
```
to be consistent with above?

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -479,19 +572,19 @@ private :
         //Note: antipodal points on equator does not define segment on equator
         //but pass by the pole
-        CT diff = geometry::math::longitude_distance_signed<geometry::radian>(lon1, lon2);
+        CT diff = geometry::math::longitude_distance_signed<geometry
+            ::radian>(lon1, lon2);
```
This line break is interesting. Are you using some kind of automatic fomatter?

vissarion replied:
could be from the previous IDE, cutting the minimal part that exceeds 80 lines.
I put it in one line now

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -527,40 +621,34 @@ private :
-                return res23;
-            }
+            
+            auto res13 = apply(lon1, lat1, lon1, half_pi * sign_non_zero, lon3, 
+                lat3, spheroid);
```
Yet another version of function call formatting

vissarion replied:
one line now

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -569,14 +657,15 @@ private :
-            boost::tuple<CT,CT> minmax_elem = boost::minmax(lat1, lat2);
-            if (lat3 >= minmax_elem.template get<0>() &&
-                lat3 <= minmax_elem.template get<1>())
+            std::tuple<CT,CT> minmax_elem = std::minmax(lat1, lat2);
```
`std::minmax()` returns `std::pair<>`. There is no need to convert it to `std::tuple<>`.

barendgehrels replied:
you can use `auto const` (and it will be a pair)

awulkiew commented on include/boost/geometry/strategies/geographic/distance_cross_track.hpp:
```diff
@@ -669,21 +760,25 @@ private :
         {
-            bisection<CT>(lon1, lat1, lon2, lat2, lon3, lat3,
-                          spheroid, res12.distance/2, res12.azimuth, result);
+            bisection(lon1, lat1, lon2, lat2, lon3, lat3, spheroid, 
+                res12.distance/2, res12.azimuth, result);
```
Ok so I see that in this file the formatting is consistent and it's the last version. I prefer the one that was here used before the change but this is ok too. The question is which one should we use in general?

vissarion replied:
So the question is align below `(` or change line and use tab, right?

barendgehrels replied:
We do both in practice. In the past I often used a tab (= spaces) but many people and IDE's now align at the `(`

awulkiew commented on include/boost/geometry/strategies/spherical/closest_points_pt_seg.hpp:
```diff
@@ -0,0 +1,200 @@
+            auto distance = distance::detail::compute_cross_track_distance::apply(
+                d_crs_pair.first, d1);
+
+            CT lon1 = geometry::get_as_radian<0>(sp1);
+            CT lat1 = geometry::get_as_radian<1>(sp1);
+            CT lon2 = geometry::get_as_radian<0>(sp2);
+            CT lat2 = geometry::get_as_radian<1>(sp2);
+
+            CT dist = CT(2) * asin(math::sqrt(distance)) * m_strategy.radius();
+            CT dist_d1 = CT(2) * asin(math::sqrt(d1)) * m_strategy.radius();
+            
+            // Note: this is similar to spherical computation in geographic
+            // point_segment_distance formula
+            CT earth_radius = m_strategy.radius();
+            CT cos_frac = cos(dist_d1 / earth_radius) / cos(dist / earth_radius);
+            CT s14_sph = cos_frac >= 1 
+                ? CT(0) : cos_frac <= -1 ? math::pi<CT>() * earth_radius
+                                         : acos(cos_frac) * earth_radius;
+
+            CT a12 = geometry::formula::spherical_azimuth<>(lon1, lat1, lon2, lat2);
```
`const`

awulkiew commented on include/boost/geometry/strategies/spherical/distance_cross_track.hpp:
```diff
@@ -50,6 +50,82 @@ namespace boost { namespace geometry
+                                 PointOfSegment const& sp1, 
+                                 PointOfSegment const& sp2)
+        {            
+            CalculationType lon1 = geometry::get_as_radian<0>(sp1);
+            CalculationType lat1 = geometry::get_as_radian<1>(sp1);
+            CalculationType lon2 = geometry::get_as_radian<0>(sp2);
+            CalculationType lat2 = geometry::get_as_radian<1>(sp2);
+            CalculationType lon = geometry::get_as_radian<0>(p);
+            CalculationType lat = geometry::get_as_radian<1>(p);
+
+            CalculationType crs_AD = geometry::formula::spherical_azimuth
```
`const`

awulkiew commented on test/algorithms/closest_points/empty_geometry.hpp:
```diff
@@ -0,0 +1,267 @@
+{
+    bg::model::segment<Point> out_seg;
+
+    try
+    {
+        bg::closest_points(geometry1, geometry2, out_seg);
+    }
+    catch(bg::empty_input_exception const& )
+    {
+        return;
+    }
+    BOOST_CHECK_MESSAGE(false,
+                        "A empty_input_exception should have been thrown");
```
Shouldn't exceptions be checked with `BOOST_CHECK_EXCEPTION`?

vissarion replied:
probably yes, but in the whole library they are checked like this so I just follow that.

awulkiew commented:
@vissarion Thanks for the PR! I apologize for the late review.

The vast majority of issues I mentioned are minor issues like formatting. The most important I think is the coordinate type of the result of pt-seg strategy which AFAIU was also merged with cartesian strategies before? Right now the coordinate type is the result type of distance strategy. I'm not sure if that's correct.

barendgehrels commented on include/boost/geometry/strategies/closest_points/spherical.hpp:
```diff
@@ -0,0 +1,112 @@
+//#include <boost/geometry/strategies/cartesian/distance_pythagoras.hpp>
+//#include <boost/geometry/strategies/spherical/distance_pythagoras_box_box.hpp>
+//#include <boost/geometry/strategies/spherical/distance_pythagoras_point_box.hpp>
+//#include <boost/geometry/strategies/spherical/distance_segment_box.hpp>
```
Can we remove these commented includes?

vissarion commented:
Thanks for the reviews! I addressed all the comments. The only issue remaining is the promotion to fp in the case of pt-seg strategy (discussion above).

awulkiew commented:
@vissarion Since this is built on top of https://github.com/boostorg/geometry/pull/933 do I have to review the other PR as well?

vissarion commented:
> @vissarion Since this is built on top of #933 do I have to review the other PR as well?

Now, you have already reviewed here.

addy90 commented:
I just noticed that in the version info on the website for Boost 1.79.0 the info about this PR and the new closest_points algorithm is missing: https://www.boost.org/users/history/version_1_79_0.html
As I follow this feature since more than a year already and also helped in finding a small bug [here](https://github.com/boostorg/geometry/pull/707#issuecomment-786650747) I would allow myself to suggest that this huge and incredible work from @vissarion should indeed be listed in the version docs on the website for the changes in boost::geometry.
This is just my opinion and I hope it is still possible to add the kudos. Also thanks again from my side for finalizing this feature in Boost 1.79.0!

awulkiew commented:
@addy90 Thanks for bringing it up. Of course it should be listed. I must have missed it.

It is possible to add it on the website. It is however not possible to add it in the documentation of geometry ver 1.79 (https://www.boost.org/doc/libs/1_79_0/libs/geometry/doc/html/geometry/release_notes.html). But it will be there in the future versions.

awulkiew commented:
@addy90 @vissarion 
https://github.com/boostorg/website/pull/681
https://github.com/boostorg/geometry/commit/e2496e2fedbc0c4468427cfb973a94afee896ca7
